app/src/main/java/io/appium/settings/LocationTracker.java

```diff
@@ -190,29 +228,58 @@ private void stopLocationUpdatesWithoutPlayServices() {
             return;
         }
 
-        Log.i(TAG, "Stopping Android location provider");
+        Log.d(TAG, "Stopping Android location provider");
         mLocationManager.removeUpdates(this);
         mLocationManager = null;
     }
-    
+
+    private boolean isPlayServicesConnected() {
+        return mGoogleApiClient != null && mGoogleApiClient.isConnected();
+    }
+
+    private boolean isLocationManagerConnected() {
+        return mLocationManager != null && mLocationProvider != null;
+    }
+
     @Nullable
-    public Location getLocation() {
-        if (mLocation == null) {
-            if (mGoogleApiClient != null) {
-                if (mGoogleApiClient.isConnected()) {
-                    try {
-                        //noinspection deprecation
-                        mLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleApiClient);
-                    } catch (SecurityException e) {
-                        Log.e(TAG, "Appium Settings has no access to location permission");
-                    }
+    public synchronized Location getLocation(Context context) {
+        if (mLocation != null) {
+            return mLocation;
+        }
+
+        // Make sure the service has been started
+        final boolean forceStart = hasAccessToLocationServices != null && !hasAccessToLocationServices;
+        start(context, GOOGLE_API_CONNECT_TIMEOUT, forceStart);
+
+        if (isPlayServicesConnected()) {
+            // Make sure the fallback is removed after play services connection succeeds
+            if (isLocationManagerConnected()) {
+                stopLocationUpdatesWithoutPlayServices();
+            }
+            try {
+                //noinspection deprecation
+                mLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleApiClient);
+                hasAccessToLocationServices = true;
+            } catch (SecurityException e) {
+                hasAccessToLocationServices = false;
+                Log.e(TAG, "Appium Settings has no access to location permission", e);
+            }
+        } else if (isLocationManagerConnected() || mGoogleApiClient != null) {
+            // Try fallback to the default provider if Google API is available but is still not connected
+            if (mGoogleApiClient != null && !isLocationManagerConnected()) {
+                mLocationManager = (LocationManager) context.getSystemService(LOCATION_SERVICE);
+                if (mLocationManager != null) {
+                    startLocationUpdatesWithoutPlayServices();
                 }
-            } else if (mLocationManager != null && locationProvider != null) {
+            }
+            if (mLocationManager != null) {
```

Comment:
Last thing. `mLocationProvider` could be null here. Should we check `mLocationProvider != null`, too?

Reply:
`isLocationManagerConnected()`? (In L267, it passed because of `mGoogleApiClient`)